app/uk/gov/hmrc/tai/controllers/income/IncomeController.scala

```diff
@@ -48,17 +53,62 @@ class IncomeController @Inject()(incomeService: IncomeService,
       } recoverWith taxAccountErrorHandler
   }
 
-  def taxCodeIncomesForYear(nino: Nino, year:TaxYear): Action[AnyContent] = authentication.async {
+  def taxCodeIncomesForYear(nino: Nino, year: TaxYear): Action[AnyContent] = authentication.async {
     implicit request =>
       incomeService.taxCodeIncomes(nino, year).map {
         case Seq() => NotFound
         case taxCodeIncomes => Ok(Json.toJson(ApiResponse(Json.toJson(taxCodeIncomes)(Writes.seq(taxCodeIncomeSourceWrites)), Nil)))
-      }recoverWith taxAccountErrorHandler
+      } recoverWith taxAccountErrorHandler
+  }
+
+  def matchedTaxCodeIncomesForYear(nino: Nino, year: TaxYear, incomeType: TaxCodeIncomeComponentType, status: TaxCodeIncomeStatus):
+  Action[AnyContent] = authentication.async {
+    implicit request =>
+
+      def filterMatchingEmploymentsToIncomeSource(employments: Seq[Employment], filteredTaxCodeIncomes: Seq[TaxCodeIncome]): Seq[IncomeSource] =
+        filteredTaxCodeIncomes.flatMap { income =>
+          employments.flatMap(emp =>
+            income.employmentId match {
+              case Some(id) if id == emp.sequenceNumber =>  Seq(IncomeSource(income, emp))
+              case _ => Seq.empty[IncomeSource]
+            }
+          )
+        }
+
+      def filterTaxCodeIncomes(taxCodeIncomes: Seq[TaxCodeIncome], incomeStatus: String): Seq[TaxCodeIncome] =
+          taxCodeIncomes.filter(income => income.componentType.toString == incomeType.toString)
+
```

Comment:
Do you need to toString these for the equality check?

Reply:
no, shouldn't do, updated